cql3/restrictions/statement_restrictions.cc

```diff
@@ -569,17 +562,514 @@ dht::partition_range_vector statement_restrictions::get_partition_key_ranges(con
+        } else if (auto dv = dynamic_pointer_cast<lists::delayed_value>(binop.rhs)) {
+            must_be_in_operator(binop);
+            process_in_values(
+                    dv->get_elements() | transformed(
+                            [&] (const ::shared_ptr<term>& t) {
+                                return static_pointer_cast<tuples::value>(t->bind(options))->get_elements();
+                            }));
+        } else if (auto mkr = dynamic_pointer_cast<tuples::in_marker>(binop.rhs)) {
+            // This is `(a,b) IN ?`.  RHS elements are themselves tuples, represented as vector<bytes_opt>.
+            must_be_in_operator(binop);
+            process_in_values(
+                    static_pointer_cast<tuples::in_value>(mkr->bind(options))->get_split_values());
+        }
+    }
+
+    void operator()(const conjunction& c) {
+        std::ranges::for_each(c.children, [this] (const expression& child) { std::visit(*this, child); });
+    }
+
+    void operator()(bool b) {
+        if (!b) {
+            ranges.clear();
+        }
+    }
+
+    static void must_be_in_operator(const binary_operator& binop) {
+        if (binop.op != oper_t::IN) {
+            on_internal_error(rlogger, format("multi_column_range_accumulator: unexpected atom {}", binop));
+        }
+    }
+
+    /// Intersects each range with v.  If any intersection is empty, clears ranges.
+    void intersect_all(const query::clustering_range& v) {
+        for (auto& r : ranges) {
+            auto intrs = intersection(r, v, prefix3cmp);
+            if (!intrs) {
+                ranges.clear();
+                break;
+            }
+            r = *intrs;
+        }
+    }
+
+    template<std::ranges::range Range>
+    requires std::convertible_to<typename Range::value_type::value_type, bytes_opt>
+    void process_in_values(Range in_values) {
+        std::vector<query::clustering_range> new_ranges;
+        std::set<std::vector<bytes_opt>> tuples_seen;
+        for (const auto& current_tuple : in_values) {
+            if (tuples_seen.contains(current_tuple)) {
+                continue; // Storage proxy doesn't accept repeated identical ranges.
+            }
+            tuples_seen.insert(current_tuple);
+            // Each IN value is like a separate EQ restriction ANDed to the existing state.
+            auto current_range = to_range(
+                    oper_t::EQ, clustering_key_prefix::from_optional_exploded(*schema, current_tuple));
+            for (const auto& r : ranges) {
+                auto intrs = intersection(r, current_range, prefix3cmp);
+                if (intrs) {
+                    new_ranges.push_back(*intrs);
+                }
+            }
+        }
+        ranges = new_ranges;
+    }
+
+    static bytes deref_column_value(bytes_opt val, sstring_view name) {
+        return *statements::request_validations::check_not_null(
+                val, "Invalid null value in condition for column %s", name);
```

Comment:
Some std::move()s here can reduce copies.

Reply:
I inlined this function; no copies should be happening here.